pkg/koordlet/util/perf_group/perf_group_linux.go

```diff
@@ -0,0 +1,409 @@
+				cgroupFile.Fd(), uintptr(cpu), uintptr(defaultFd), uintptr(unix.PERF_FLAG_PID_CGROUP|unix.PERF_FLAG_FD_CLOEXEC), uintptr(0))
+			if e1 != syscall.Errno(0) {
+				errMutex.Lock()
+				err = multierr.Append(err, fmt.Errorf("failed to create perf fd, Error: %s, cpu: %d, event: %s", unix.ErrnoName(e1), cpu, events[0]))
+				errMutex.Unlock()
+				return
+			}
+			leaderFd := os.NewFile(r1, fmt.Sprintf("%s_%d", events[0], cpu))
+			pc.leaderFd = leaderFd
+			var id uint64
+			_, _, e1 = collector.syscall6(syscall.SYS_IOCTL, r1, uintptr(unix.PERF_EVENT_IOC_ID), uintptr(unsafe.Pointer(&id)), 0, 0, 0)
+			if e1 != syscall.Errno(0) {
+				errMutex.Lock()
+				err = multierr.Append(err, fmt.Errorf("failed to get perf id, Error: %s, cpu: %d, event: %s", unix.ErrnoName(e1), cpu, events[0]))
+				errMutex.Unlock()
+				return
+			}
+			collector.mu.Lock()
+			collector.idEventMap[id] = events[0]
+			collector.mu.Unlock()
+			for i := 1; i < len(events); i++ {
+				attr := attrMap[events[i]]
+				r1, _, e1 := collector.syscall6(syscall.SYS_PERF_EVENT_OPEN, uintptr(unsafe.Pointer(attr)),
+					cgroupFile.Fd(), uintptr(cpu), r1, uintptr(unix.PERF_FLAG_PID_CGROUP|unix.PERF_FLAG_FD_CLOEXEC), uintptr(0))
+				if e1 != syscall.Errno(0) {
+					errMutex.Lock()
+					err = multierr.Append(err, fmt.Errorf("failed to create perf fd, Error: %s, cpu: %d, event: %s", unix.ErrnoName(e1), cpu, events[i]))
+					errMutex.Unlock()
+					return
+				}
+				fd := os.NewFile(r1, fmt.Sprintf("%s_%d", events[i], cpu))
+				pc.fds = append(pc.fds, fd)
+				var id uint64
+				_, _, e1 = collector.syscall6(syscall.SYS_IOCTL, r1, uintptr(unix.PERF_EVENT_IOC_ID), uintptr(unsafe.Pointer(&id)), 0, 0, 0)
+				if e1 != syscall.Errno(0) {
+					errMutex.Lock()
+					err = multierr.Append(err, fmt.Errorf("failed to get perf id, Error: %s, cpu: %d, event: %s", unix.ErrnoName(e1), cpu, events[i]))
+					errMutex.Unlock()
+					return
+				}
+				collector.mu.Lock()
+				collector.idEventMap[id] = events[i]
+				collector.mu.Unlock()
+			}
+			// enable perf group
+			_, _, e1 = collector.syscall6(syscall.SYS_IOCTL, leaderFd.Fd(), uintptr(unix.PERF_EVENT_IOC_RESET), unix.PERF_IOC_FLAG_GROUP, 0, 0, 0)
+			if e1 != syscall.Errno(0) {
+				errMutex.Lock()
+				err = multierr.Append(err, fmt.Errorf("failed to reset perf group, Error: %s, cpu: %d, events: %s", unix.ErrnoName(e1), cpu, events))
+				errMutex.Unlock()
+				return
+			}
+			_, _, e1 = collector.syscall6(syscall.SYS_IOCTL, leaderFd.Fd(), uintptr(unix.PERF_EVENT_IOC_ENABLE), unix.PERF_IOC_FLAG_GROUP, 0, 0, 0)
+			if e1 != syscall.Errno(0) {
+				errMutex.Lock()
+				err = multierr.Append(err, fmt.Errorf("failed to enable perf group, Error: %s, cpu: %d, events: %s", unix.ErrnoName(e1), cpu, events))
+				errMutex.Unlock()
+				return
+			}
+			collector.perfCollectors.Store(cpu, &pc)
+		}(cpu)
+	}
+	wg.Wait()
+	for _, attr := range attrMap {
+		C.free(unsafe.Pointer(attr))
```

Comment:
is the memory leak possible if return at L162 with error

Reply:
Fix with `defer C.free(attr)` after create attr.

Reply:
> is the memory leak possible if return at L162 with error

Fix with defer C.free(attr) after create attr.